rfcs/experimental/flow_graph_msg_keys/README.md

```diff
@@ -0,0 +1,226 @@
+# Type-specified message keys for join_node
+
+## Introduction
+
+This experimental extension allows a key matching `join_node` to obtain keys via functions associated with its
+input types. The extension simplifies the existing approach by removing the need to provide a function object for
+each input port of `join_node`.
+
+Let's consider the following graph that has two `queue_nodes` named `q0` and `q1` that are connected to a
+key matching `join_node j` that is then connected to a final `queue_node` named `q3`:
+
+<img src="JoinGraph.png">
+
+The code to implement such a graph is shown below, with the message type `Message`.
+
+```cpp
+#include "oneapi/tbb/flow_graph.h"
+#include <iostream>
+
+namespace msg {
+    template<typename K>
+    struct Message {
+        K k;
+        int v;
+    };
+
+    template <typename K>
+    std::ostream& operator<<(std::ostream& os, const Message<K>& m) {
+        os << "(" << m.k << ", " << m.v << ")";
+        return os;
+    }
+}
+
+int main(int argc, char *argv[]) {
+    using msg_t = msg::Message<int>;
+    tbb::flow::graph g;
+
+    tbb::flow::queue_node<msg_t> q0{g};
+    tbb::flow::queue_node<msg_t> q1{g};
+
+    tbb::flow::join_node<std::tuple<msg_t, msg_t>, tbb::flow::key_matching<int>> 
+    j{g, [](const msg_t& m) { return m.k; }, [](const msg_t& m) { return m.k; } };
+
+    tbb::flow::queue_node<std::tuple<msg_t,msg_t>> q3{g};
+
+    tbb::flow::make_edge(q0, tbb::flow::input_port<0>(j));
+    tbb::flow::make_edge(q1, tbb::flow::input_port<1>(j));
+    tbb::flow::make_edge(j, q3);
+
+    int n = 10;
+    for (int i = 1; i < n; ++i) {
+       q0.try_put(msg_t{i, 1000+i});
+       q1.try_put(msg_t{n-i, 10000+i});
+    }
+    g.wait_for_all();
+
+    for (int i = 1; i < n; ++i) {
+        std::tuple<msg_t, msg_t> t;
+        q3.try_get(t);
+        std::cout << std::get<0>(t) << ":" << std::get<1>(t) << std::endl;
+    }
+    return 0;
+}
+
+```
+
+The key part to note in the example is the definition of `join_node j`:
+
+```cpp
+    tbb::flow::join_node<std::tuple<msg_t, msg_t>, tbb::flow::key_matching<int>> 
+    j{g, [](const msg_t& m) { return m.k; }, [](const msg_t& m) { return m.k; } };
+```
+
+For each input port, a function object must be provided to get the key from the incoming message.
+For this example, there are two identical lambda expressions passed, `[](const msg_t& m) { return m.k; }`.
+If a `join_node` has 10 ports, there will need to be 10 function objects passed to the constructor.
+
+This proposal simplifies the use of key matching `join_node` by providing a way to associate a
```

Comment:
Since this document describes the existing feature, I would avoid using the term "proposal":)

Reply:
Fixed.